src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+/// for free as long as you adhere to the requirements described in the
+/// comments below.
+pub trait CollapsePrivate: Clone + Send + Sized + Sized {
+    // *********************************************************** //
+    // ********************* REQUIRED METHODS ******************** //
+    // *********************************************************** //
+
+    /// Some formats, such as `dtrace`, contain a header or other non-stack
+    /// information at the beginning of their input files. If header information
+    /// is present, this method **must** consume it (i.e. advance the provided
+    /// reader past it).
+    ///
+    /// This method also provides an opportunity to do processing of actual
+    /// stack data on the main thread before worker threads are spun up. For
+    /// an example of why this might be necessary, see `perf`, which can require
+    /// reading the first stack in order to know how to process the rest.
+    ///
+    /// If the format you are working with does not contain header information
+    /// or does not need any special, up-front processing, just have this method
+    /// return `Ok(())` immediately.
+    fn pre_process<R>(&mut self, reader: &mut R, occurrences: &mut Occurrences) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The primary method.
+    ///
+    /// It receives a reader whose header has already been consumed (see above), as
+    /// well as a mutable reference to an `Occurences` instance (just a hashamp that
+    /// works across multiple threads). Implementators should parse the stack data
+    /// contained in the reader and write output to the provided `Occurrences` map.
+    ///
+    /// So that this method may be called multiple times, all internal
+    /// state contained in `self` (e.g. any stack buffers or the like) **must** be
+    /// reset before this method returns. Also, this method may **not** use threads.
+    fn collapse_single_threaded<R>(
+        &mut self,
+        reader: R,
+        occurrences: &mut Occurrences,
+    ) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// This method receives a line from the input data (including the trailing newline byte)
+    /// and should return `true` if the line represents the end of a stack; `false` otherwise.
+    fn is_end_of_stack(&self, line: &[u8]) -> bool;
+
+    /// This method should return whether the implementation corresponds with
+    /// the given input string, i.e. if the input data matches the collapser.
+    ///
+    /// - `None` means "not sure -- need more input"
+    /// - `Some(true)` means "yes, this implementation should work with this string"
+    /// - `Some(false)` means "no, this implementation definitely won't work"
+    fn is_applicable_(&mut self, input: &str) -> Option<bool>;
+
+    /// This method should return the number of stacks per job to send to the threadpool.
+    fn nstacks_per_job(&self) -> usize;
```

Comment:
It's weird to me for this to be a method on `CollapsePrivate`. I'd assume it was a property of the multi-threaded "wrapper" instead?

Reply:
Same with the next three methods.

Reply:
I'm not sure I'm understanding your comment. `nthreads` and `nstacks_per_job`, if we we **not** trying to be generic, would just be simple fields on a struct. Because we need to access those "fields" in `collapse_multithreaded`, however, we need some way of accessing "fields" even though we're dealing with a trait, not a concrete struct. The simplest way to do this is to have the implementer expose the "fields" through getter/setter methods.

Reply:
Actually, the getters are there because we need to use them in `collapse_multithreaded'; the setters, on the other hand, are there so we can use them in the tests. Just thought I'd clarify.

Reply:
Yeah, I didn't give you too much context here :sweat_smile: My thinking was that you could have a `MultithreadedCollapser` wrapper struct:
```rust
struct MultithreadedCollapse<C: CollapsePrivate> {
  nthreads: usize,
  collapser: C,
}
```
and then implement `Collapse` for that wrapper struct rather than for `T where T: Collapse`. I'm not sure it would be worth it in terms of complicating the higher-level code though, so probably just stick with what we have. It just makes me sad to have to repeat the `nthreads` option in both the `dtrace` and `perf` modules.